app/slave/cluster_slave.py

```diff
@@ -173,18 +175,20 @@ def _do_build_teardown_and_reset(self, timeout=None):
         for executor in self.executors_by_id.values():
             executor.kill()
 
-        if not self._project_type:
-            return  # There is no build to tear down! (Slave is idle.)
+        # This lock keeps multiple teardown attempts from racing each other and throwing exceptions.
+        with self._build_teardown_lock:
```

Comment:
@tjlee0909 I don't think it's completely redundant -- I think the coin was meant as more of an assertion on behavior than as a concurrency control mechanism. But I think you're right that if we are adding a stronger control like a lock, we probably won't need the weaker check provided by the coin.

@drobertduke, under what conditions were you encountering the exception, and was it the `BuildTeardownError` from the coin or was it something else?

At some point (when we added the coin) there was clearly an assumption that we were not going to call this more than once for a single build. The coin was essentially an assertion for that assumption. It doesn't seem like that assumption has changed, but it seems like we now need extra concurrency control around this logic and so maybe the coin will be obsolete.

I don't disagree that a lock might be necessary (depending on which exceptions you were seeing) but one alternate approach to think about is only locking around setting `self._project_type` to `None`. That way other threads would not have to wait for teardown to complete:
```python
with self._build_teardown_lock:
    if not self._project_type:
        return
    project_type_to_teardown = self._project_type
    self._project_type = None

project_type_to_teardown.teardown_build(timeout=timeout)
...
```